hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/metrics/MetricsImpl.java

```diff
@@ -16,55 +16,51 @@
 
 package com.hazelcast.jet.impl.metrics;
 
-import com.hazelcast.jet.core.metrics.Counter;
-import com.hazelcast.jet.core.metrics.Gauge;
+import com.hazelcast.jet.core.metrics.Metric;
 import com.hazelcast.jet.core.metrics.Unit;
-import com.hazelcast.jet.impl.execution.ProcessorTasklet;
 import com.hazelcast.jet.impl.execution.Tasklet;
 
-public final class UserMetricsImpl {
+import javax.annotation.Nullable;
+
+public final class MetricsImpl {
 
     private static final ThreadLocal<Container> CONTEXT = ThreadLocal.withInitial(Container::new);
 
-    private UserMetricsImpl() {
+    private MetricsImpl() {
     }
 
-    public static Container getContainer() {
+    public static Container container() {
         return CONTEXT.get();
     }
 
-    public static Counter getCounter(String name) {
-        return getContext().getCounter(name);
-    }
-
-    public static Gauge getGauge(String name, Unit unit) {
-        return getContext().getGauge(name, unit);
+    public static Metric metric(String name, Unit unit, boolean threadSafe) {
+        return getContext().metric(name, unit, threadSafe);
     }
 
-    private static UserMetricsContext getContext() {
+    private static MetricsContext getContext() {
         Container container = CONTEXT.get();
-        UserMetricsContext context = container.getContext();
+        MetricsContext context = container.getContext();
         if (context == null) {
-            throw new RuntimeException("User metrics are accessible only from internal worker threads");
+            throw new RuntimeException("Thread %s has no user-metrics context set; this is a bug only " +
+                    "if it is an internal Jet thread; otherwise user-metrics related code needs to be " +
+                    "modified so that it is not explicitly run on foreign threads");
         }
         return context;
     }
 
     public static class Container {
 
-        private UserMetricsContext context;
+        private MetricsContext context;
 
         Container() {
         }
 
-        public UserMetricsContext getContext() {
+        public MetricsContext getContext() {
             return context;
         }
 
-        public void setContext(Tasklet tasklet) {
-            if (tasklet instanceof ProcessorTasklet) {
-                this.context = ((ProcessorTasklet) tasklet).getUserMetricsContext();
-            }
+        public void setContext(@Nullable Tasklet tasklet) {
```

Comment:
wouldn't it be better to pass the context directly here instead from the ExecutionService?

Reply:
Not sure I understand, pls. rephrase.

Reply:
Set the call to `MetricsImpl.container().setContext(t.metricsContext())` rather than passing `Tasklet`